### PR TITLE
8308396: Fix offset_of conversion warnings in runtime code

### PR DESCRIPTION
**Pull request by coleenp**
Please review this change to use ByteSize and byte_offset_of() to refer to offsets to metadata and other types that are used in generated code.  This avoids the int narrowing conversion warning for -Wconversion.  There were a couple that I just added an (int) cast instead because these offsets are either being used in other code currently being changed (in oopDesc) or there are too many (like displaced_header_offset_in_bytes) and should be their own change.

This change takes a  chunk out of the -Wconversion warnings - see CR for more info.

It might be easier and less tedious to review the commits separately.  One commit renames blah_offset_in_bytes to blah_offset, since in_bytes(blah_offset()) is typically used (except in Address constructor which has an overload for ByteSize).

Tested with tier1-4, x86 and aarch64, and built linux-x64-zero linux-x64-zero-debug linux-aarch64-debug linux-s390x-open linux-arm32-debug linux-ppc64le-debug linux-riscv64-debug locally.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308396](https://bugs.openjdk.org/browse/JDK-8308396): Fix offset_of conversion warnings in runtime code

### Reviewers
 * [Amit Kumar](https://openjdk.org/census#amitkumar) (@offamitkumar - Author) ⚠️ Review applies to [56f2b4ba](https://git.openjdk.org/jdk/pull/14053/files/56f2b4ba54544211e46930a97584a8fb9e1db657)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [f629ebce](https://git.openjdk.org/jdk/pull/14053/files/f629ebcee098cfb8aac5401ff39aca4d264771f7)
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**) ⚠️ Review applies to [1f78a6fe](https://git.openjdk.org/jdk/pull/14053/files/1f78a6fe8b8ba72471355d4b141d7b693fe96503)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14053/head:pull/14053` \
`$ git checkout pull/14053`

Update a local copy of the PR: \
`$ git checkout pull/14053` \
`$ git pull https://git.openjdk.org/jdk.git pull/14053/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14053`

View PR using the GUI difftool: \
`$ git pr show -t 14053`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14053.diff">https://git.openjdk.org/jdk/pull/14053.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14053#issuecomment-1553750758)

### CODE REVIEW

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp [hunk 1]
```diff
@@ -2778,7 +2778,7 @@ void SharedRuntime::generate_uncommon_trap_blob() {
 #ifdef ASSERT
   assert(Immediate::is_uimm8(Deoptimization::Unpack_LIMIT), "Code not fit for larger immediates");
   assert(Immediate::is_uimm8(Deoptimization::Unpack_uncommon_trap), "Code not fit for larger immediates");
-  const int unpack_kind_byte_offset = Deoptimization::UnrollBlock::unpack_kind_offset_in_bytes()
+  const int unpack_kind_byte_offset = Deoptimization::UnrollBlock::unpack_kind_offset()
 #ifndef VM_LITTLE_ENDIAN
   + 3
 #endif
```
Comment (offamitkumar) [inline src/hotspot/cpu/s390/sharedRuntime_s390.cpp:2784]: This is breaking build for s390x. 

```
/home/amit/jdk/src/hotspot/cpu/s390/sharedRuntime_s390.cpp: In static member function 'static void SharedRuntime::generate_uncommon_trap_blob()':           
/home/amit/jdk/src/hotspot/cpu/s390/sharedRuntime_s390.cpp:2783:3: error: no match for 'operator+' (operand types are 'ByteSize' and 'int')                 
   const int unpack_kind_byte_offset = Deoptimization::UnrollBlock::unpack_kind_offset()                                                                    
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                    
 #ifndef VM_LITTLE_ENDIAN                                                                                                                                   
 ~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                   
   + 3                                                                                                                                                      
   ^~~                                                                                                                                                      
In file included from /home/amit/jdk/src/hotspot/share/utilities/exceptions.hpp:31,                                                                         
                 from /home/amit/jdk/src/hotspot/share/oops/metadata.hpp:28,                                                                                
                 from /home/amit/jdk/src/hotspot/share/oops/oop.hpp:32,                                                                                     
                 from /home/amit/jdk/src/hotspot/share/runtime/handles.hpp:29,
                 from /home/amit/jdk/src/hotspot/share/code/oopRecorder.hpp:28,                                                                            
                 from /home/amit/jdk/src/hotspot/share/asm/codeBuffer.hpp:28,
                 from /home/amit/jdk/src/hotspot/share/asm/assembler.hpp:28,
                 from /home/amit/jdk/src/hotspot/share/asm/macroAssembler.hpp:28,                                                                          
                 from /home/amit/jdk/src/hotspot/share/asm/macroAssembler.inline.hpp:28,                                                                   
                 from /home/amit/jdk/src/hotspot/cpu/s390/sharedRuntime_s390.cpp:27:                                                                       
/home/amit/jdk/src/hotspot/share/utilities/sizes.hpp:53:20: note: candidate: 'constexpr ByteSize operator+(ByteSize, ByteSize)'                            
 constexpr ByteSize operator + (ByteSize x, ByteSize y) { return in_ByteSize(in_bytes(x) + in_bytes(y)); }                                                 
                    ^~~~~~~~
/home/amit/jdk/src/hotspot/share/utilities/sizes.hpp:53:20: note:   no known conversion for argument 2 from 'int' to 'ByteSize'                            
gmake[3]: *** [lib/CompileJvm.gmk:147: /home/amit/jdk/build/linux-s390x-server-fastdebug/hotspot/variant-server/libjvm/objs/sharedRuntime_s390.o] Error 1  
gmake[3]: *** Waiting for unfinished jobs....
gmake[2]: *** [make/Main.gmk:252: hotspot-server-libs] Error 2

ERROR: Build failed for target 'images' in configuration 'linux-s390x-server-fastdebug' (exit code 2)                                                      
Stopping javac server
```

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp [hunk 2]
```diff
@@ -2778,7 +2778,7 @@ void SharedRuntime::generate_uncommon_trap_blob() {
 #ifdef ASSERT
   assert(Immediate::is_uimm8(Deoptimization::Unpack_LIMIT), "Code not fit for larger immediates");
   assert(Immediate::is_uimm8(Deoptimization::Unpack_uncommon_trap), "Code not fit for larger immediates");
-  const int unpack_kind_byte_offset = Deoptimization::UnrollBlock::unpack_kind_offset_in_bytes()
+  const int unpack_kind_byte_offset = Deoptimization::UnrollBlock::unpack_kind_offset()
```
Comment (offamitkumar) [inline src/hotspot/cpu/s390/sharedRuntime_s390.cpp:2781]:
```suggestion
  const int unpack_kind_byte_offset = in_bytes(Deoptimization::UnrollBlock::unpack_kind_offset())
```

**Review by offamitkumar — APPROVED** [review]
Testing for fastdebug build/ tier1 tests looks good on s390x.

I've given my suggestion for fixing the build-break, but I leave that upto you :-)

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: Fixed.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 3]
```diff
@@ -4311,7 +4310,7 @@ void MacroAssembler::load_method_holder_cld(Register rresult, Register rmethod)
 void MacroAssembler::load_method_holder(Register holder, Register method) {
   ldr(holder, Address(method, Method::const_offset()));                      // ConstMethod*
   ldr(holder, Address(holder, ConstMethod::constants_offset()));             // ConstantPool*
-  ldr(holder, Address(holder, ConstantPool::pool_holder_offset_in_bytes())); // InstanceKlass*
+  ldr(holder, Address(holder, ConstantPool::pool_holder_offset())); // InstanceKlass*
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:4313]: Nit: Alignment.

**Review by jdksjolen — APPROVED** [review]
Hi,

This looks good to me, with one small nit.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: Thanks for reviewing!  Fixed.

File: src/hotspot/cpu/s390/sharedRuntime_s390.cpp (hunk 1 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/sharedRuntime_s390.cpp]: oh, Got it. 

BTW thanks for considering suggestion & for this change as well 🙂

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 4]
```diff
@@ -5116,7 +5116,7 @@ void MacroAssembler::load_method_holder_cld(Register rresult, Register rmethod)
 void MacroAssembler::load_method_holder(Register holder, Register method) {
   movptr(holder, Address(method, Method::const_offset()));                      // ConstMethod*
   movptr(holder, Address(holder, ConstMethod::constants_offset()));             // ConstantPool*
-  movptr(holder, Address(holder, ConstantPool::pool_holder_offset_in_bytes())); // InstanceKlass*
+  movptr(holder, Address(holder, ConstantPool::pool_holder_offset())); // InstanceKlass*
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp]: Thanks - I fixed one but not two others that were  no longer aligned.  Now fixed, with these other changes.

**Comment by coleenp** [comment]
Thank you for reviewing Amit, Johan and Fred.
/integrate

**Pull request closed without merging**